ddspipe_participants/src/cpp/participant/dds/CommonParticipant.cpp

```diff
@@ -0,0 +1,349 @@
+            this->payload_pool_,
+            dds_participant_,
+            fastdds_topic,
+            discovery_database_);
+        reader->init();
+
+        return reader;
+    }
+    else
+    {
+        auto reader = std::make_shared<SimpleReader>(
+            this->id(),
+            dds_topic,
+            this->payload_pool_,
+            dds_participant_,
+            fastdds_topic);
+        reader->init();
+
+        return reader;
+    }
+}
+
+void CommonParticipant::on_subscriber_discovery(
+        fastdds::dds::DomainParticipant* ,
+        fastrtps::rtps::ReaderDiscoveryInfo&& info)
+{
+    // If reader is from other participant, store it in discovery database
+    if (detail::come_from_same_participant_(info.info.guid(), this->dds_participant_->guid()))
+    {
+        // Come from this participant, do nothing
+        return;
+    }
+
+    // Calculate endpoint info
+    core::types::Endpoint info_reader =
+        detail::create_endpoint_from_info_<fastrtps::rtps::ReaderDiscoveryInfo>(info, id());
+
+    // If new endpoint discovered
+    if (info.status == fastrtps::rtps::ReaderDiscoveryInfo::DISCOVERY_STATUS::DISCOVERED_READER)
+    {
+        // TODO check logic because if an endpoint is lost by liveliness it may be inserted again when already in database
+        this->discovery_database_->add_endpoint(info_reader);
+    }
+    else
+    {
+        this->discovery_database_->update_endpoint(info_reader);
+    }
+}
+
+void CommonParticipant::on_publisher_discovery(
+        fastdds::dds::DomainParticipant* ,
+        fastrtps::rtps::WriterDiscoveryInfo&& info)
+{
+    // If writer is from other participant, store it in discovery database
+    if (detail::come_from_same_participant_(info.info.guid(), this->dds_participant_->guid()))
+    {
+        // Come from this participant, do nothing
+        return;
+    }
+
+    // Calculate endpoint info
+    core::types::Endpoint info_writer =
+        detail::create_endpoint_from_info_<fastrtps::rtps::WriterDiscoveryInfo>(info, id());
+
+    // If new endpoint discovered
+    if (info.status == fastrtps::rtps::WriterDiscoveryInfo::DISCOVERY_STATUS::DISCOVERED_WRITER)
+    {
+        // TODO check logic because if an endpoint is lost by liveliness it may be inserted again when already in database
+        this->discovery_database_->add_endpoint(info_writer);
+    }
+    else
+    {
+        this->discovery_database_->update_endpoint(info_writer);
+    }
+}
+
+CommonParticipant::CommonParticipant(
+        const std::shared_ptr<SimpleParticipantConfiguration>& participant_configuration,
+        const std::shared_ptr<core::PayloadPool>& payload_pool,
+        const std::shared_ptr<core::DiscoveryDatabase>& discovery_database)
+    : configuration_(participant_configuration)
+    , payload_pool_(payload_pool)
+    , discovery_database_(discovery_database)
+{
+    // Do nothing
+}
+
+fastdds::dds::DomainParticipantQos CommonParticipant::reckon_participant_qos_() const
```

Comment:
As already discussed offline, it may be useful to reuse this function (additionally accepting a participant configuration as parameter), to configure transport for example.

This could become a public static function, be in the new utils section within pipe, or even put it somewhere in dev-utils.

Reply:
future feature for when is needed

Reply:
TODO in future PR